### DISCUSSION TITLE
Suddenly HA Sensors Not Updating Properly

### DISCUSSION THREAD

**Discussion by 503Dev** [tsightler/ring-mqtt#506] [Support Q&A] [answered]
####Description:
Suddenly the sensors in HA specifically "Motion" and "Ding" no longer report the correct status.  When motion is detected or a ding the sensor simply shows as "Unavailable" and then goes back to "Clear".  Notifications via the Ring mobile app are working properly, battery levels full on doorbell and additional sensors like battery level, etc. update just fine.

####Environment: 
Latest version of HA, Mosquito MQTT addon and ring add on running on a linux server with supervised installation of HA.
Ring camera is a gen 1 wireless doorbell.

####Diagnostic Info:
Checking the logs via HA for the Ring-MQTT plugin does not show any errors.  Regular updates of status, authentication token gets updated when it should, etc.  Connectivity to the doorbell is normal and I can force a live stream, etc.  Screenshots do not update on motion however this appears to be related to the actual doorbell when no ring protect plan is purchased, otherwise everything works normally.

####Diagnostic MQTT Inspection:
Inspecting the MQTT topic with MQTT Explorer I can see the ring information as expected.  In addition the values for last updated / last detected for motion and ding are timestamps which correspond correctly to the events I expect.  This seems to indicate the doorbell is working normally and the updates do get pushed to the MQTT stack however HA sensors show incorrect values.

####Troubleshooting Attempted:
-Disabled any automations using the camera sensors and live stream
-Verified live stream is off and notifications via Ring app work as expected
-Removed, reinstalled and logged in with new authentication to verify connectivity
-Restarted HA and everything else just to be sure

**Comment by tsightler** [comment]
Logs from a ding or motion event are the best way to troubleshoot, but, realistically, it seems unlikely to be an issue with ring-mqtt if it's just suddenly stopped working and the MQTT status looks correct.  Did you look at the HA logs to see if there's anything showing up there?

**Reply by 503Dev** [reply]
I did and not seeing anything obvious.  I will pull logs shortly and post them here, I am honestly stumped as to where the failure is at right now.  I am a programmer myself so I may be overthinking something obvious.

**Reply by tsightler** [reply]
The symptom sounds like the service is crashing, but I don't know why that would be, especially not "suddenly".  The fact that it's somehow going "Unavailable" implies that the ding is being detected by ring-mqtt, so it's up to whether the state is sending the correct value or not. 
 Did you perhaps recently end your subscription, or have you always used ring-mqtt this way?

**Reply by 503Dev** [reply]
I have always used ring the same, I have not had a sub for a year.  It has worked fine until recently.  I just uncovered some new symptoms, I restarted my server and when the Ring MQTT plugin came back up it threw an error that the saved refresh token does not appear valid.

Per the usual steps I clicked "Open the Web UI" and put in my credentials.  I hit submit and nothing at all happens.  Pulling the dev tools  Network tab shows an outbound request _submit-account_ however it appears to fail.  It does not get a response.  Pulling the URL on that request I can see:

`http://192.168.68.190:8123/api/hassio_ingress/3Dx6_flYzwecQFhFDL4vXNpl1bzr68ChuVUBTnUPPW4/submit-account`

That is the correct IP/port for my local server but if I attempt to open that URL I get:

`Error: Cannot GET /submit-account`

As a result I cannot advance now to renew the refresh token. Ring-MQTT logs do not show anything aside from the usual refresh token expired error.

**Reply by 503Dev** [reply]
Ok so another reboot and it reconnected, no idea why.  Anyway, after it reconnected the Ring-MQTT in HA shows everything normal in logs.  However the status for motion and ding is still incorrect.

However, if I inspect the MQTT topics #ring/<device id>/ camera/<id>/motion and /ding they both have available attributes and inside those I can see the lastMotion value has the correct timestamp yet in HA it does not propagate properly.

From MQTT Explorer: 
`
{
  "lastMotion": 1669272315,
  "lastMotionTime": "2022-11-24T06:45:15Z",
  "personDetected": false,
  "motionDetectionEnabled": true
}`

That is correct, 12:45AM my local time however looking at Home Assistant the device -> sensors -> ding and motion show clear and then random unavailable right whenever activity occurs.  They never show detected anymore.  I just ran a quick test and triggered motion on the doorbell and looked at the HA Ring-MQTT log, it does not appear to crash but it does not register anything.  However shortly after via MQTT-Explorer the ring topic updates.

**Reply by tsightler** [reply]
The more important value is the state topic vs the attribute topic.  It should only ever be "ON" or "OFF", any other value is invalid and it should be updated whenever a new motion or ding is detected to "ON" and then, back to "OFF" once the event is expired.  You should see new push notification being received or the expire time logged just before the state changes.  Does the entire camera device go into offline/unavailable state or just the motion/ding entity itself?  It's a really weird issues, I'm struggling to understand how it could be.

I will say that I think Ring had some issues with their push notifications yesterday, both my production and development systems struggled with registering for push events for several hours, even though the Ring status page shows no issues.

**Reply by 503Dev** [reply]
I also agree with the idea that ring had issues.  I noticed when attempting to visit the website at one point it returned a weird error.

The ON or OFF state always sits as OFF.  The device does not appear to go offline.

Even though push notifications seemed to break yesterday with the ring app I did notice a workaround (not relevant to this plugin) but if you sign in with a browser to ring.com and enable push notifications in the browser those continue to work even when API / app notifications get stuck.

**Comment by 503Dev** [comment, marked as answer]
Just wanted to provide an update.  This can be closed I suppose, I do not think the issue is with the Ring-MQTT plugin (thank you for all your awesome work btw).

I think this an issue with Ring service which they fail to recognize.  I can confirm that if I login to "www.ring.com" sometimes it will give me a "406 http response" which would seem like Ring has blocked my IP but debugging traffic suggests more likely some type of weird routing issue or a routing/load balancer issue on their end.  I search their forums and as usual they try to blame the users and their ISP (I work in IT and I am already sure my ISP is not to blame this time).

Some people complain this happens with a VPN but I think more likely it has to do with the way traffic is being routed on certain blocks of addresses or regions.  It appears to affect the way Ring-MQTT gets responses too or just misses them entirely even though the Ring-MQTT -> Doorbell communication is fine as I get info updates with no issue.

Tldr; Ring's API for notifications is defective and does not work right, it even fails with their own app sometimes.

**Reply by tsightler** [reply]
I was getting 429 responses for a while last night and failures to register the push receiver on restarts, lasted about an hour or so and then things recovered.  Ring actually has a lot of these "minor" outages that never show up on the status page.  The alarm API seems pretty solid, but the cameras, they are pretty horrible, although, actually, much improved over a few years ago.

**Reply by 503Dev** [reply]
Thanks for the update. Yes indeed that appears to be it. I setup some monitoring and it looks like this happens very regularly and they do not report it. Your implementation appears pretty solid but alas reliability depends on ring.

**Reply by tsightler** [reply]
I do worry that somehow the implementation is triggering the throttling, but I would think it would be far less random if that were the case.  For example, the native HA integration is notorious for triggering throttling, however, my setup can run months without any throttling and then suddenly will see these 429 responses multiple times in a few days.

**Reply by 503Dev** [reply]
I also agree, I did not suspect your implementation specifically because I literally can go months and months with 0 issues.  Also there can be 1-2 hour periods in the early AM where it fails but it's fine in the morning.  This seems much more like issues with the load balancing on the Ring DC servers than your implementation.  Plus I noticed a funny quirk - if I appear to be throttled on ring.com I can just visit any of the regional sub domains such as https://latam-es.ring.com/ and it works fine. If there was an IP throttle or block it would be logical sub domains would be blocked too.
